privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,578 @@
+
+        We need to define the function again, to get rid of the
+        is_challenge_request-decorator of the base class
+
+        :param passw: password, which might be pin or pin+otp
+        :param options: dictionary of additional request parameters
+
+        :return: returns true or false
+        """
+        return self.check_pin(passw, user=user, options=options)
+
+    def create_challenge(self, transactionid=None, options=None):
+        """
+        This method creates a challenge, which is submitted to the user.
+        The submitted challenge will be preserved in the challenge
+        database.
+
+        If no transaction id is given, the system will create a transaction
+        id and return it, so that the response can refer to this transaction.
+
+        :param transactionid: the id of this challenge
+        :param options: the request context parameters / data
+        :type options: dict
+        :return: tuple of (bool, message, transactionid, attributes)
+        :rtype: tuple
+
+        The return tuple builds up like this:
+        ``bool`` if submit was successful;
+        ``message`` which is displayed in the JSON response;
+        additional ``attributes``, which are displayed in the JSON response.
+        """
+        options = options or {}
+        message = get_action_values_from_options(SCOPE.AUTH,
+                                                 ACTION.CHALLENGETEXT,
+                                                 options) or DEFAULT_CHALLENGE_TEXT
+
+        attributes = None
+        data = None
+        challenge = b32encode(geturandom())
+        fb_identifier = self.get_tokeninfo(PUSH_ACTION.FIREBASE_CONFIG)
+        if fb_identifier:
+            # We send the challenge to the Firebase service
+            fb_gateway = create_sms_instance(fb_identifier)
+            url = fb_gateway.smsgateway.option_dict.get(FIREBASE_CONFIG.REGISTRATION_URL)
+            message_on_mobile = get_action_values_from_options(SCOPE.AUTH,
+                                                   PUSH_ACTION.MOBILE_TITLE,
+                                                   options) or DEFAULT_MOBILE_TEXT
+            title = get_action_values_from_options(SCOPE.AUTH,
+                                                   PUSH_ACTION.MOBILE_TITLE,
+                                                   options) or "privacyIDEA"
+            smartphone_data = {"nonce": challenge,
+                               "question": message_on_mobile,
+                               "serial": self.token.serial,
+                               "title": title,
+                               "url": url}
+            # Create the signature.
+            # value to string
+            sign_string = u"{nonce}|{url}|{serial}|{question}|{title}".format(**smartphone_data)
+
+            pem_privkey = self.get_tokeninfo(PRIVATE_KEY_SERVER)
+            privkey_obj = serialization.load_pem_private_key(str(pem_privkey), None, default_backend())
```

Comment:
We need to use ``to_bytes`` instead of ``str`` for Python 3 compatibility here.

Reply:
done in c3f36915